core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1272,8 +1272,8 @@ protected Engine engineUnsafe() {
         @Override
         public void onFailedEngine(ShardId shardId, String reason, @Nullable Throwable failure) {
             try {
-                // delete the shard state so this folder will not be reused
-                MetaDataStateFormat.deleteMetaState(nodeEnv.availableShardPaths(shardId));
+                // mark as corrupted, so opening the store will fail
```

Comment:
note how the Throwable may be null. Also, it would be great if we can capture the reason. So I would suggest we make an IOException here with a message that has the reason in it and capture the failure as a cause.